bot/src/cogs/shop.py

```diff
@@ -0,0 +1,149 @@
+
+if TYPE_CHECKING:
+    from src.wrapper.schema import ShopItem
+
+
+async def item_autocomplete(ctx: AutocompleteContext) -> list[str]:
+    shop = (
+        ctx.bot.interface.shop.iter_items()  # pyright: ignore[reportAttributeAccessIssue]
+    )
+
+    return [item.name async for item in shop]
+
+
+class Confirm(discord.ui.View):
+    """Discord view with confirm/cancel buttons."""
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.confirmed = None
+
+    @discord.ui.button(label="Confirm", style=discord.ButtonStyle.green)
+    async def confirm_callback(
+        self,
+        _,  # noqa: ANN001
+        interaction: discord.Interaction,
+    ) -> None:
+        """Ran when confirm button is pressed."""
+        await interaction.response.defer()
+        self.confirmed = True
+        self.stop()
+
+    @discord.ui.button(label="Cancel", style=discord.ButtonStyle.grey)
+    async def cancel_callback(
+        self,
+        _,  # noqa: ANN001
+        interaction: discord.Interaction,
+    ) -> None:
+        """Ran when cancel button is pressed."""
+        await interaction.response.defer()
+        self.confirmed = False
+        self.stop()
+
+
+class Shop(commands.Cog):
+    """Cog holding everything related to the shop."""
+
+    def __init__(self, client: Client) -> None:
+        self.client = client
+
+    shop = SlashCommandGroup("shop")
+
+    @shop.command()
+    async def view(self, ctx: Context) -> None:
+        """Interact with the shop."""
+        # Temporary while wrapper is getting fixed
+        shop = self.client.interface.shop.iter_items()
+
+        shop_pages: list[discord.Embed] = []
+
+        # Yes, I know about enumerate but apparently it doesnt work with async generators
+        index = 0
+        async for item in shop:
+            if index % 6 == 0:
+                shop_pages.append(
+                    discord.Embed(
+                        title="Shop",
+                        description="`/shop buy <name>` to buy an item",
+                    )
+                )
+
+            disabled_tag = "~~" if item.quantity == 0 else ""
+
+            shop_pages[-1].add_field(
+                name=f"{disabled_tag}{item.name}{disabled_tag}",
+                value=f"${item.price:.2f}\n{item.quantity} in stock",
+            )
+            index += 1
+
+        paginator = pages.Paginator(
+            pages=shop_pages  # pyright: ignore[reportArgumentType]
+        )
+        await paginator.respond(ctx.interaction)
+
+    @shop.command()
+    @option(
+        "item",
+        str,
+        description="The item you want to buy.",
+        autocomplete=item_autocomplete,
+    )
+    @option("quantity", int, description="How many you want to buy, defaults to 1.")
+    async def buy(self, ctx: Context, item: str, quantity: int = 1) -> None:
+        """Buy an item in the shop."""
+        try:
+            company = await self.client.interface.company.get_company(ctx.author.id)
```

Comment:
Doesn't need to get the company, you can just give the string user id as `company` in purchase

Reply:
I get the company to make sure they have one, and if they don't I can return an error message